app/javascript/topNavigation/utilities.js

```diff
@@ -1,107 +1,103 @@
-  if (crayonsHeaderMenuClassList.contains('showing')) {
-    crayonsHeaderMenuClassList.remove('showing');
-    navigationButton.setAttribute('aria-expanded', 'false');
-  } else {
-    crayonsHeaderMenuClassList.add('showing');
-    navigationButton.setAttribute('aria-expanded', 'true');
-
-    const firstNavLink = document.getElementById('first-nav-link');
-    if (firstNavLink) {
-      setTimeout(() => {
-        // focus first item on open
-        firstNavLink.focus();
-      }, 100);
+  // Focus on the first item in the menu
+  (function focusFirstItem() {
+    if (document.activeElement === firstItem) {
+      // The first element has focus
+      return;
     }
-  }
+
+    firstItem.focus();
+    // requestAnimationFrame is faster and more reliable than setTimeout
+    // https://swizec.com/blog/how-to-wait-for-dom-elements-to-show-up-in-modern-browsers
+    window.requestAnimationFrame(focusFirstItem);
+  })();
 }
 
+/**
+ * Determines whether or not a device is a touch device.
+ *
+ * @returns true if a touch device, otherwise false.
+ */
 export function isTouchDevice() {
   return /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini|DEV-Native-ios/i.test(
     navigator.userAgent,
   );
 }
 
-export function initializeTouchDevice(memberTopMenu, menuNavButton) {
+/**
+ * Initializes the member navigation menu events.
+ *
+ * @param {HTMLElement} memberTopMenu The member menu in the top navigation.
+ * @param {HTMLElement} menuNavButton The button to activate the member navigation menu.
+ */
+export function initializeMemberMenu(memberTopMenu, menuNavButton) {
+  // Typically using CSS for hovering for the menu is the way to go. But... since we use InstantClick for
+  // loading pages, the top header navigation never changes in terms of the DOM references.
+  // Because of this, we're using mouse events to mouseover/mouseout on the member's avatar
+  // to attach styles to get it to show the menu so that this works on desktop and mobile.
   if (navigator.userAgent === 'DEV-Native-ios') {
     document.body.classList.add('dev-ios-native-body');
   }
 
-  if (memberTopMenu) {
-    const crayonsHeaderMenuClassList = memberTopMenu.classList;
+  menuNavButton.addEventListener('click', (_event) => {
+    if (memberTopMenu.classList.contains('showing')) {
+      closeHeaderMenu(memberTopMenu, menuNavButton);
+    } else {
+      openHeaderMenu(memberTopMenu, menuNavButton);
+    }
+  });
 
-    setTimeout(() => {
+  const { classList } = memberTopMenu;
+
+  if (isTouchDevice()) {
+    memberTopMenu.addEventListener('focus', (_event) => {
+      menuNavButton.setAttribute('aria-expanded', 'true');
+    });
+  } else {
+    memberTopMenu.addEventListener('mouseover', (_event) => {
+      classList.add('desktop');
+      openHeaderMenu(memberTopMenu, menuNavButton);
+    });
+    memberTopMenu.addEventListener('mouseout', (_event) => {
       closeHeaderMenu(memberTopMenu, menuNavButton);
+    });
```

Comment:
One thing I noticed while testing this on desktop, is that if I open the menu with the keyboard, then move my mouse so that it moves over and away from the menu, it collapses the menu again:

https://user-images.githubusercontent.com/20773163/115363488-6d058780-a1ba-11eb-9d0b-e9586f05e739.mp4

When this happens, the menuNavButton isn't refocused on close. I think probably we only want to close the menu on 'mouseout' if it wasn't opened by keyboard... or at a minimum make sure the the menuNavButton is refocused when it does close.

I also noticed that I can no longer click the menuNavButton on desktop and have the menu stay visible until I click an option or away from it (only works on hover now). Are we comfortable losing that feature? I'm not sure I have a strong opinion on it, but it is different than the previous functionality.

Reply:
> One thing I noticed while testing this on desktop, is that if I open the menu with the keyboard, then move my mouse so that it moves over and away from the menu, it collapses the menu again:
> 
> 
> 
> https://user-images.githubusercontent.com/20773163/115363488-6d058780-a1ba-11eb-9d0b-e9586f05e739.mp4
> 
> 
> 
> When this happens, the menuNavButton isn't refocused on close. I think probably we only want to close the menu on 'mouseout' if it wasn't opened by keyboard... or at a minimum make sure the the menuNavButton is refocused when it does close.
> 
> 
> 
> I also noticed that I can no longer click the menuNavButton on desktop and have the menu stay visible until I click an option or away from it (only works on hover now). Are we comfortable losing that feature? I'm not sure I have a strong opinion on it, but it is different than the previous functionality.

Definitely not comfortable losing that feature. I'll take a peek at this too.

Reply:
This has been fixed. I actually noticed that in production we do not focus on the menu nav button after clicking on a link in the menu or losing focus on the menu.